k2/python/tests/mutual_information_test.py

```diff
@@ -0,0 +1,271 @@
+#!/usr/bin/env python3
+#
+# Copyright      2021  Xiaomi Corporation   (authors: Daniel Povey,
+#                                                     Wei Kang)
+#
+# See ../../../LICENSE for clarification regarding multiple authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# To run this single test, use
+#
+#  ctest --verbose -R mutual_information_test_py
+
+import random
+import unittest
+
+import k2
+import torch
+
+
+# Caution: this will fail occasionally due to cutoffs not being quite large
+# enough. As long as it passes most of the time, it's OK.
+class TestMutualInformation(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        cls.devices = [torch.device('cpu')]
+        if torch.cuda.is_available() and k2.with_cuda:
+            cls.devices.append(torch.device('cuda', 0))
+            if torch.cuda.device_count() > 1:
+                torch.cuda.set_device(1)
+                cls.devices.append(torch.device('cuda', 1))
+        cls.dtypes = [torch.float32, torch.float64]
+
+    def test_mutual_information_basic(self):
+        for _iter in range(100):
+            (B, S, T) = (random.randint(1, 10), random.randint(1, 16),
+                         random.randint(1, 500))
+            random_px = (random.random() < 0.2)
+            random_py = (random.random() < 0.2)
+            random_boundary = (random.random() < 0.7)
+            big_px = (random.random() < 0.2)
+            big_py = (random.random() < 0.2)
+
+            for dtype in self.dtypes:
+                for device in self.devices:
+                    if random_boundary:
+
+                        def get_boundary_row():
+                            s_begin = random.randint(0, S - 1)
+                            t_begin = random.randint(0, T - 1)
+                            # allow empty sequence
+                            s_end = random.randint(s_begin, S)
+                            # allow empty sequence
+                            t_end = random.randint(t_begin, T)
+                            return [s_begin, t_begin, s_end, t_end]
+
+                        if device == torch.device('cpu'):
+                            boundary = torch.tensor(
+                                [get_boundary_row() for _ in range(B)],
+                                dtype=torch.int64,
+                                device=device)
+                        else:
+                            boundary = boundary.to(device)
```

Comment:
You are using the `boundary` that was defined when `device` is `cpu`.

The code should work if users disable the tests for CPU.